adagucserverEC/CDataReader.cpp

```diff
@@ -530,33 +611,65 @@ int CDataReader::parseDimensions(CDataSource *dataSource,int mode,int x, int y,
       
     }
     
-    status = dataSource->varX->readData(CDF_DOUBLE,sta,sto,str);
+    status = dataSource->varX->readData(CDF_DOUBLE,sta,sto,str,true);
     if(status!=0){
       CDBError("Unable to read x dimension with name %s for variable %s",dataSource->varX->name.c_str(),dataSourceVar->name.c_str());
       return 1;
     }
-    
+
 //     CDBDebug("Read data X %f",((double*)dataSource->varX->data)[0]);
-    
+
+
     //CDBDebug("Done");
     
     sta[0]=start[dataSource->dimYIndex];str[0]=dataSource->stride2DMap; sto[0]=dataSource->dHeight;
     if(singleCellMode){
       sta[0]=0;str[0]=1;sto[0]=2;
-      
     }
     
-    status = dataSource->varY->readData(CDF_DOUBLE,sta,sto,str);if(status!=0){
+    status = dataSource->varY->readData(CDF_DOUBLE,sta,sto,str,true);
+    if(status!=0){
       CDBError("Unable to read y dimension for variable %s",dataSourceVar->name.c_str());
       for(size_t j=0;j<dataSource->varY->dimensionlinks.size();j++){
             CDBDebug("For var %s, reading dim %s of size %d (%d %d %d)", dataSource->varY->name.c_str(),dataSource->varY->dimensionlinks[j]->name.c_str(),dataSource->varY->dimensionlinks[j]->getSize(),sta[j],sto[j],str[j]);
           }
       return 1;
     }
+
+    // Get CRS info DataObject
+    getCRS(dataSource);
+
+    //if units="rad" and grid_mapping variable contains perspective_point_height attribute
+    //coordinates can be converted to (k)m
+    //
+    CDF::Attribute* units=dataSource->varX->getAttributeNE("units");
+    CDBDebug("units: %s", units->toString().c_str());
+    if ((units!=NULL) && units->toString().equals("rad")){
+      CDBDebug("Correct varX and varY");
+      double *xdata=(double*)dataSource->varX->data;
+      double *ydata=(double*)dataSource->varY->data;
+      CDF::Attribute *projvarnameAttr = dataSourceVar->getAttributeNE("grid_mapping");
+      if(projvarnameAttr!=NULL){
+	CDF::Variable * projVar = cdfObject->getVariableNE(projvarnameAttr->toString().c_str());
+	if(projVar==NULL){CDBWarning("projection variable '%s' not found",(char*)projvarnameAttr->data);}
+	else {
+	  //Get perspective_height
+	  CDF::Attribute *perspectiveHeightAttr = projVar->getAttributeNE("perspective_point_height");
+	  double perspectiveHeight=35786000.;
+          if (perspectiveHeightAttr!=NULL) {
+	  try{perspectiveHeight = perspectiveHeightAttr->toString().toDouble();}catch(int e){}
+          }
+
+	  for (size_t j=sta[0];j<sto[0];j+=str[0]) {
+	    xdata[j]=xdata[j]*perspectiveHeight;
+	    ydata[j]=ydata[j]*perspectiveHeight;
```

Comment:
Loop individually with right array sizes, sometimes arrays do not have same length